plugins/formal-verification/formver.uniqueness/src/org/jetbrains/kotlin/formver/UniqueCFA.kt

```diff
@@ -0,0 +1,171 @@
+                            setOf(UniqueLevel.Unique)
+                        } else {
+                            setOf(UniqueLevel.Shared)
+                        }
+                    )
+                }.toMap()
+            return dataForNode.transformValues { it.putAll(valueParameters) }
+        }
+
+        override fun visitVariableDeclarationNode(
+            node: VariableDeclarationNode,
+            data: PathAwareControlFlowInfo<FirVariableSymbol<FirVariable>, Set<UniqueLevel>>,
+        ): PathAwareControlFlowInfo<FirVariableSymbol<FirVariable>, Set<UniqueLevel>> {
+            val dataForNode = visitNode(node, data)
+            val lSymbol = node.fir.symbol
+            if (node.fir.initializer == null) {
+                return dataForNode.transformValues { it.put(lSymbol, setOf(UniqueLevel.Shared)) }
+            }
+            val rhsUnique = when (val last = context.uniqueStack.last().last()) {
+                is Level -> last.level
+                is Path -> dataForNode[NormalPath]?.get(last.symbol) ?: setOf(UniqueLevel.Shared)
+            }
+            return dataForNode.transformValues { it.put(lSymbol, rhsUnique) }
+        }
+
+        override fun visitVariableAssignmentNode(
+            node: VariableAssignmentNode,
+            data: PathAwareControlFlowInfo<FirVariableSymbol<FirVariable>, Set<UniqueLevel>>,
+        ): PathAwareControlFlowInfo<FirVariableSymbol<FirVariable>, Set<UniqueLevel>> {
+            val dataForNode = visitNode(node, data)
+            val lSymbol = node.fir.lValue.toResolvedCallableSymbol(context.session)
+            val rhsUnique = when (val last = context.uniqueStack.last().last()) {
+                is Level -> last.level
+                is Path -> dataForNode[NormalPath]?.get(last.symbol) ?: setOf(UniqueLevel.Shared)
+            }
+            return dataForNode.transformValues { it.put(lSymbol as FirVariableSymbol, rhsUnique) }
+        }
+
+        @OptIn(SymbolInternals::class)
+        override fun visitFunctionCallArgumentsExitNode(
+            node: FunctionCallArgumentsExitNode,
+            data: PathAwareControlFlowInfo<FirVariableSymbol<FirVariable>, Set<UniqueLevel>>,
+        ): PathAwareControlFlowInfo<FirVariableSymbol<FirVariable>, Set<UniqueLevel>> {
+            // TODO: this is where function call parameter checking should happen
+            // Here do not call visitNode
+            val params = (node.fir.toResolvedCallableSymbol() as FirFunctionSymbol<*>).fir.valueParameters
+            val argumentAliasOrUnique: MutableList<PathUnique> = mutableListOf()
+            for (i in params.indices) {
+                argumentAliasOrUnique.add(context.uniqueStack.last().popLast())
+            }
+            // FIXME: might not get annotation in this way
+            params.zip(argumentAliasOrUnique).forEach { (param, varUnique) ->
+                val argShouldUnique = param.hasAnnotation(context.uniqueId, context.session)
```

Comment:
`should` is ambiguous when used like this.  Does this mean the parameter is declared unique?  If so, call it `paramIsUnique` or something.  (Though since there is already a `UniqueLevel` enum, `paramUniqueness: UniqueLevel` would be even better.)